icu4c/source/common/ulocbuilder.cpp

```diff
@@ -0,0 +1,173 @@
+  if (builder == nullptr) return;
+  internal(builder)->setLanguage(
+      length < 0 ? StringPiece(language) : StringPiece(language, length));
+}
+
+void ulocbld_setScript(ULocaleBuilder* builder, const char* script, int32_t length) {
+  if (builder == nullptr) return;
+  internal(builder)->setScript(
+      length < 0 ? StringPiece(script) : StringPiece(script, length));
+}
+
+void ulocbld_setRegion(ULocaleBuilder* builder, const char* region, int32_t length) {
+  if (builder == nullptr) return;
+  internal(builder)->setRegion(
+      length < 0 ? StringPiece(region) : StringPiece(region, length));
+}
+
+void ulocbld_setVariant(ULocaleBuilder* builder, const char* variant, int32_t length) {
+  if (builder == nullptr) return;
+  internal(builder)->setVariant(
+      length < 0 ? StringPiece(variant) : StringPiece(variant, length));
+}
+
+void ulocbld_setExtension(ULocaleBuilder* builder, char key, const char* value, int32_t length) {
+  if (builder == nullptr) return;
+  internal(builder)->setExtension(key,
+      length < 0 ? StringPiece(value) : StringPiece(value, length));
+}
+
+void ulocbld_setUnicodeLocaleKeyword(
+    ULocaleBuilder* builder, const char* key, int32_t keyLength,
+    const char* type, int32_t typeLength) {
+  if (builder == nullptr) return;
+  internal(builder)->setUnicodeLocaleKeyword(
+      keyLength < 0 ? StringPiece(key) : StringPiece(key, keyLength),
+      typeLength < 0 ? StringPiece(type) : StringPiece(type, typeLength));
+}
+
+void ulocbld_addUnicodeLocaleAttribute(
+    ULocaleBuilder* builder, const char* attribute, int32_t length) {
+  if (builder == nullptr) return;
+  internal(builder)->addUnicodeLocaleAttribute(
+      length < 0 ? StringPiece(attribute) : StringPiece(attribute, length));
+}
+
+void ulocbld_removeUnicodeLocaleAttribute(
+    ULocaleBuilder* builder, const char* attribute, int32_t length) {
+  if (builder == nullptr) return;
+  internal(builder)->removeUnicodeLocaleAttribute(
+      length < 0 ? StringPiece(attribute) : StringPiece(attribute, length));
+}
+
+void ulocbld_clear(ULocaleBuilder* builder) {
+  if (builder == nullptr) return;
+  internal(builder)->clear();
+}
+
+void ulocbld_clearExtensions(ULocaleBuilder* builder) {
+  if (builder == nullptr) return;
+  internal(builder)->clearExtensions();
+}
+
+int32_t ulocbld_buildLocaleID(ULocaleBuilder* builder,
+                  char* buffer, int32_t bufferCapacity, UErrorCode* err) {
+  if (builder == nullptr) {
+    *err = U_ILLEGAL_ARGUMENT_ERROR;
+    return 0;
+  }
+  icu::Locale l = internal(builder)->build(*err);
+  if (U_FAILURE(*err)) return 0;
+  int32_t length = (int32_t)(uprv_strlen(l.getName()));
+  if (length + 1 > bufferCapacity) {
+    *err = U_BUFFER_OVERFLOW_ERROR;
+    if (bufferCapacity > 0) {
+        uprv_memcpy(buffer, l.getName(), bufferCapacity-1);
+        // Always null terminate.
+        buffer[bufferCapacity - 1] = '\0';
```

Comment:
I actually like this, but didn't we have a fairly big argument in ICU TC not that long about about my proposal to make the `uloc_getXXX()` methods always null-terminate their result buffers?  I guess there's a backward-compatibility issue with doing it to those methods (although it doesn't actually seem that bad to me), but the big issue seemed to be that we've always defined all these methods that fill in buffers as _not_ null-terminating the buffer if the result's length equals or exceeds the buffer length.  Is it okay for this method to behave differently?  Do we need to re-litigate that discussion?  @markusicu

Reply:
sorry, I just have bad code here. This code will cause uprv_memcpy to segfault because bufferCapacity could be much biggger than the memory allocated for l.getName(). The code in this PR is not ready for review yet since I am still waiting for the API to be approved by the ICU TC and I have not yet complete the testing code.